File: src/attestation/signing.ts
/**
 * Ed25519 Signing and Verification for Attestations
 *
 * Implements §4.2-4.4 of the RecourseOS Attestation Protocol.
 * Uses Node.js crypto module for Ed25519 operations.
 */

import { createHash, generateKeyPairSync, sign, verify, KeyObject } from 'crypto';
import { canonicalize } from './canonicalize.js';
import { KeyMetadata, canSign, canVerify } from './key-management.js';

/**
 * Error thrown when signing or verification fails
 */
export class AttestationSigningError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'AttestationSigningError';
  }
}

/**
 * Base64url encode (RFC 4648 Section 5, no padding)
 */
export function base64urlEncode(buffer: Buffer): string {
  return buffer.toString('base64url');
}

/**
 * Base64url decode (RFC 4648 Section 5)
 */
export function base64urlDecode(str: string): Buffer {
  return Buffer.from(str, 'base64url');
}

/**
 * Attestation content fields (signed payload excludes signature and attestation_uri for ID derivation)
 */
export interface AttestationContent {
  input: unknown;
  output: unknown;
  evaluator: string;
  timestamp: string;
  key_id: string;
}

/**
 * Full attestation structure
 */
export interface Attestation extends AttestationContent {
  attestation_uri: string;
  signature: string;
}

/**
 * Generate an Ed25519 keypair
 *
 * @returns Object containing private and public key in various formats
 */
export function generateKeyPair(): {
  privateKey: KeyObject;
  publicKey: KeyObject;
  publicKeyBase64url: string;
} {
  const { privateKey, publicKey } = generateKeyPairSync('ed25519');
  const publicKeyRaw = publicKey.export({ type: 'spki', format: 'der' });
  // Ed25519 public key in SPKI format has a 12-byte header, raw key is last 32 bytes
  const rawPublicKey = publicKeyRaw.slice(-32);

  return {
    privateKey,
    publicKey,
    publicKeyBase64url: base64urlEncode(rawPublicKey),
  };
}

/**
 * Derive attestation ID from content fields per §6.3
 *
 * The attestation_id is derived from: input, output, evaluator, timestamp, key_id
 * Excludes: signature, attestation_uri (to avoid circularity)
 *
 * @param content - The attestation content fields
 * @returns 32-character lowercase hex string
 */
export function deriveAttestationId(content: AttestationContent): string {
  // Construct object with exactly these fields
  const idPayload = {
    input: content.input,
    output: content.output,
    evaluator: content.evaluator,
    timestamp: content.timestamp,
    key_id: content.key_id,
  };

  // Canonicalize per RFC 8785
  const canonical = canonicalize(idPayload);

  // SHA-256 hash
  const hash = createHash('sha256').update(canonical).digest();

  // First 16 bytes as lowercase hex (32 characters)
  return hash.slice(0, 16).toString('hex');
}

/**
 * Construct attestation URI from instance base URL and attestation ID
 */
export function constructAttestationUri(
  instanceBaseUrl: string,
  attestationId: string
): string {
  // Remove trailing slash if present
  const baseUrl = instanceBaseUrl.replace(/\/$/, '');
  return `${baseUrl}/.well-known/attestations/${attestationId}.json`;
}

/**
 * Construct the signed payload from an attestation (all fields except signature)
 *
 * Per §4.1: The signed payload consists of all attestation fields except signature.
 * The attestation_uri IS included in the signed payload.
 */
export function constructSignedPayload(
  attestation: Omit<Attestation, 'signature'>
): string {
  return canonicalize(attestation);
}

/**
 * Sign an attestation per §4.3
 *
 * @param content - The attestation content fields
 * @param privateKey - The Ed25519 private key
 * @param keyMetadata - Metadata about the signing key (must be active)
 * @param instanceBaseUrl - Base URL for constructing attestation_uri
 * @returns Complete signed attestation
 */
export function signAttestation(
  content: AttestationContent,
  privateKey: KeyObject,
  keyMetadata: KeyMetadata,
  instanceBaseUrl: string
): Attestation {
  // Verify key is in active state (§4.3, §5.2)
  if (!canSign(keyMetadata)) {
    throw new AttestationSigningError(
      `Key "${keyMetadata.key_id}" is in state "${keyMetadata.state}" and cannot sign. ` +
        'Only keys in "active" state may sign new attestations.'
    );
  }

  // Verify key_id matches
  if (content.key_id !== keyMetadata.key_id) {
    throw new AttestationSigningError(
      `Content key_id "${content.key_id}" does not match key metadata "${keyMetadata.key_id}"`
    );
  }

  // Step 1-2: Derive attestation_id from content fields
  const attestationId = deriveAttestationId(content);

  // Step 3: Construct attestation_uri
  const attestationUri = constructAttestationUri(instanceBaseUrl, attestationId);

  // Step 4: Build attestation without signature
  const unsignedAttestation: Omit<Attestation, 'signature'> = {
    ...content,
    attestation_uri: attestationUri,
  };

  // Step 5: Construct signed payload (canonicalized)
  const signedPayload = constructSignedPayload(unsignedAttestation);

  // Step 6: Sign with Ed25519
  const signatureBuffer = sign(null, Buffer.from(signedPayload), privateKey);

  // Step 7: Encode as base64url
  const signature = base64urlEncode(signatureBuffer);

  return {
    ...unsignedAttestation,
    signature,
  };
}

/**
 * Verification result
 */
export type VerificationResult =
  | { valid: true }
  | { valid: false; reason: string };

/**
 * Verify an attestation signature per §4.4
 *
 * @param attestation - The attestation to verify
 * @param publicKeyBase64url - The base64url-encoded public key
 * @param keyMetadata - Metadata about the key (must be verifiable state)
 * @returns Verification result
 */
export function verifyAttestation(
  attestation: Attestation,
  publicKeyBase64url: string,
  keyMetadata: KeyMetadata
): VerificationResult {
  // Step 4: Check key state (§4.4)
  if (!canVerify(keyMetadata)) {
    return {
      valid: false,
      reason: `Key "${keyMetadata.key_id}" is in state "${keyMetadata.state}" and cannot be used for verification. ` +
        'Keys in "pending" or "compromised" state fail verification.',
    };
  }

  // Verify key_id matches
  if (attestation.key_id !== keyMetadata.key_id) {
    return {
      valid: false,
      reason: `Attestation key_id "${attestation.key_id}" does not match provided key "${keyMetadata.key_id}"`,
    };
  }

  try {
    // Step 1: Extract and decode signature
    const signatureBuffer = base64urlDecode(attestation.signature);

    // Step 2: Construct signed payload (remove signature, canonicalize)
    const { signature: _, ...unsignedAttestation } = attestation;
    const signedPayload = constructSignedPayload(unsignedAttestation);

    // Step 3: Decode public key
    // Node.js requires SPKI format for Ed25519 public keys
    // Raw Ed25519 public key is 32 bytes, we need to wrap it in SPKI
    const rawPublicKey = base64urlDecode(publicKeyBase64url);
    if (rawPublicKey.length !== 32) {
      return {
        valid: false,
        reason: `Invalid public key length: expected 32 bytes, got ${rawPublicKey.length}`,
      };
    }

    // SPKI header for Ed25519 (12 bytes)
    const spkiHeader = Buffer.from([
      0x30, 0x2a, // SEQUENCE, 42 bytes
      0x30, 0x05, // SEQUENCE, 5 bytes (AlgorithmIdentifier)
      0x06, 0x03, // OID, 3 bytes
      0x2b, 0x65, 0x70, // 1.3.101.112 (Ed25519)
      0x03, 0x21, // BIT STRING, 33 bytes
      0x00, // unused bits = 0
    ]);
    const spkiPublicKey = Buffer.concat([spkiHeader, rawPublicKey]);

    // Step 5: Verify Ed25519 signature
    const isValid = verify(
      null,
      Buffer.from(signedPayload),
      { key: spkiPublicKey, format: 'der', type: 'spki' },
      signatureBuffer
    );

    if (!isValid) {
      return {
        valid: false,
        reason: 'Ed25519 signature verification failed',
      };
    }

    // Step 6: Verification succeeded
    return { valid: true };
  } catch (error) {
    return {
      valid: false,
      reason: `Verification error: ${error instanceof Error ? error.message : String(error)}`,
    };
  }
}

/**
 * Verify attestation ID matches content
 *
 * Use this to check that an attestation_uri corresponds to the actual content.
 */
export function verifyAttestationId(attestation: Attestation): boolean {
  const derivedId = deriveAttestationId(attestation);
  // Extract ID from attestation_uri
  const match = attestation.attestation_uri.match(/\/([a-f0-9]{32})\.json$/);
  if (!match) {
    return false;
  }
  return match[1] === derivedId;
}

/**
 * Create a complete attestation (convenience function)
 *
 * This combines content creation, ID derivation, URI construction, and signing.
 */
export function createAttestation(
  input: unknown,
  output: unknown,
  evaluator: string,
  privateKey: KeyObject,
  keyMetadata: KeyMetadata,
  instanceBaseUrl: string
): Attestation {
  const content: AttestationContent = {
    input,
    output,
    evaluator,
    timestamp: new Date().toISOString(),
    key_id: keyMetadata.key_id,
  };

  return signAttestation(content, privateKey, keyMetadata, instanceBaseUrl);
}

File: recourse-go/pkg/attestation/signing.go
// Package attestation implements cryptographic attestation for consequence reports.
//
// This implements the RecourseOS Attestation Protocol, enabling cross-implementation
// verification between Go and TypeScript implementations.
package attestation

import (
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"regexp"
	"strings"
	"time"
)

// Attestation represents a signed consequence evaluation attestation.
// This structure matches the TypeScript implementation for cross-implementation interop.
type Attestation struct {
	Input          interface{} `json:"input"`
	Output         interface{} `json:"output"`
	Evaluator      string      `json:"evaluator"`
	Timestamp      string      `json:"timestamp"`
	KeyID          string      `json:"key_id"`
	AttestationURI string      `json:"attestation_uri"`
	Signature      string      `json:"signature"` // Ed25519, base64url encoded
}

// Signer handles attestation signing operations.
type Signer struct {
	privateKey      ed25519.PrivateKey
	publicKey       ed25519.PublicKey
	keyID           string
	evaluator       string
	instanceBaseURL string // Base URL for attestation URIs
}

// NewSigner creates a new Signer with a fresh Ed25519 key pair.
// The instanceBaseURL is used to construct attestation URIs (e.g., "https://example.com").
func NewSigner(evaluator, instanceBaseURL string) (*Signer, error) {
	pub, priv, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("failed to generate key pair: %w", err)
	}

	// Key ID is first 8 bytes of SHA-256 of public key, hex encoded
	hash := sha256.Sum256(pub)
	keyID := hex.EncodeToString(hash[:8])

	return &Signer{
		privateKey:      priv,
		publicKey:       pub,
		keyID:           keyID,
		evaluator:       evaluator,
		instanceBaseURL: strings.TrimSuffix(instanceBaseURL, "/"),
	}, nil
}

// NewSignerFromKey creates a Signer from an existing private key.
// The key should be 64 bytes (Ed25519 private key).
func NewSignerFromKey(evaluator, instanceBaseURL string, privateKey []byte) (*Signer, error) {
	if len(privateKey) != ed25519.PrivateKeySize {
		return nil, fmt.Errorf("invalid private key size: expected %d, got %d", ed25519.PrivateKeySize, len(privateKey))
	}

	priv := ed25519.PrivateKey(privateKey)
	pub := priv.Public().(ed25519.PublicKey)

	hash := sha256.Sum256(pub)
	keyID := hex.EncodeToString(hash[:8])

	return &Signer{
		privateKey:      priv,
		publicKey:       pub,
		keyID:           keyID,
		evaluator:       evaluator,
		instanceBaseURL: strings.TrimSuffix(instanceBaseURL, "/"),
	}, nil
}

// PublicKey returns the public key bytes.
func (s *Signer) PublicKey() []byte {
	return s.publicKey
}

// PrivateKey returns the private key bytes.
func (s *Signer) PrivateKey() []byte {
	return s.privateKey
}

// KeyID returns the key identifier.
func (s *Signer) KeyID() string {
	return s.keyID
}

// Sign creates an attestation for the given input and output.
// This follows the RecourseOS Attestation Protocol:
// 1. Derive attestation_id from content fields (excluding attestation_uri and signature)
// 2. Construct attestation_uri from instance base URL
// 3. Sign the full attestation (including attestation_uri, excluding signature)
func (s *Signer) Sign(input, output interface{}) (*Attestation, error) {
	timestamp := time.Now().UTC().Format(time.RFC3339)

	// Step 1: Build content payload for ID derivation (excludes attestation_uri and signature)
	idPayload := map[string]interface{}{
		"input":     input,
		"output":    output,
		"evaluator": s.evaluator,
		"timestamp": timestamp,
		"key_id":    s.keyID,
	}

	// Canonicalize for ID derivation
	idCanonical, err := Canonicalize(idPayload)
	if err != nil {
		return nil, fmt.Errorf("failed to canonicalize ID payload: %w", err)
	}

	// Compute attestation ID: SHA-256(canonical).slice(0, 16).hex()
	hash := sha256.Sum256([]byte(idCanonical))
	attestationID := hex.EncodeToString(hash[:16])

	// Step 2: Construct attestation URI (URL format for TypeScript compatibility)
	attestationURI := fmt.Sprintf("%s/.well-known/attestations/%s.json", s.instanceBaseURL, attestationID)

	// Step 3: Build signed payload (includes attestation_uri, excludes signature)
	signedPayload := map[string]interface{}{
		"attestation_uri": attestationURI,
		"evaluator":       s.evaluator,
		"input":           input,
		"key_id":          s.keyID,
		"output":          output,
		"timestamp":       timestamp,
	}

	// Canonicalize the signed payload
	signedCanonical, err := Canonicalize(signedPayload)
	if err != nil {
		return nil, fmt.Errorf("failed to canonicalize signed payload: %w", err)
	}

	// Sign the canonical payload
	signature := ed25519.Sign(s.privateKey, []byte(signedCanonical))
	signatureB64 := base64.RawURLEncoding.EncodeToString(signature)

	return &Attestation{
		Input:          input,
		Output:         output,
		Evaluator:      s.evaluator,
		Timestamp:      timestamp,
		KeyID:          s.keyID,
		AttestationURI: attestationURI,
		Signature:      signatureB64,
	}, nil
}

// Verifier handles attestation verification operations.
type Verifier struct {
	publicKeys map[string]ed25519.PublicKey // keyID -> publicKey
}

// NewVerifier creates a new Verifier.
func NewVerifier() *Verifier {
	return &Verifier{
		publicKeys: make(map[string]ed25519.PublicKey),
	}
}

// AddPublicKey registers a public key for verification.
func (v *Verifier) AddPublicKey(keyID string, publicKey []byte) error {
	if len(publicKey) != ed25519.PublicKeySize {
		return fmt.Errorf("invalid public key size: expected %d, got %d", ed25519.PublicKeySize, len(publicKey))
	}
	v.publicKeys[keyID] = ed25519.PublicKey(publicKey)
	return nil
}

// Verify checks the attestation signature.
// Returns nil if valid, error if invalid.
func (v *Verifier) Verify(attestation *Attestation) error {
	// Look up the public key
	publicKey, ok := v.publicKeys[attestation.KeyID]
	if !ok {
		return fmt.Errorf("unknown key ID: %s", attestation.KeyID)
	}

	// Reconstruct the signed payload (includes attestation_uri, excludes signature)
	signedPayload := map[string]interface{}{
		"attestation_uri": attestation.AttestationURI,
		"evaluator":       attestation.Evaluator,
		"input":           attestation.Input,
		"key_id":          attestation.KeyID,
		"output":          attestation.Output,
		"timestamp":       attestation.Timestamp,
	}

	// Canonicalize the signed payload
	signedCanonical, err := Canonicalize(signedPayload)
	if err != nil {
		return fmt.Errorf("failed to canonicalize signed payload: %w", err)
	}

	// Decode the signature
	signature, err := base64.RawURLEncoding.DecodeString(attestation.Signature)
	if err != nil {
		return fmt.Errorf("failed to decode signature: %w", err)
	}

	// Verify the signature
	if !ed25519.Verify(publicKey, []byte(signedCanonical), signature) {
		return fmt.Errorf("invalid signature")
	}

	// Verify the attestation ID matches the content
	idPayload := map[string]interface{}{
		"input":     attestation.Input,
		"output":    attestation.Output,
		"evaluator": attestation.Evaluator,
		"timestamp": attestation.Timestamp,
		"key_id":    attestation.KeyID,
	}

	idCanonical, err := Canonicalize(idPayload)
	if err != nil {
		return fmt.Errorf("failed to canonicalize ID payload: %w", err)
	}

	hash := sha256.Sum256([]byte(idCanonical))
	expectedID := hex.EncodeToString(hash[:16])

	// Extract attestation ID from URI and verify
	actualID, err := ExtractAttestationID(attestation.AttestationURI)
	if err != nil {
		return fmt.Errorf("failed to extract attestation ID: %w", err)
	}

	if actualID != expectedID {
		return fmt.Errorf("attestation ID mismatch: expected %s, got %s", expectedID, actualID)
	}

	return nil
}

// VerifyWithPublicKey verifies an attestation with a provided public key,
// without requiring pre-registration.
func VerifyWithPublicKey(attestation *Attestation, publicKey []byte) error {
	if len(publicKey) != ed25519.PublicKeySize {
		return fmt.Errorf("invalid public key size: expected %d, got %d", ed25519.PublicKeySize, len(publicKey))
	}

	// Verify the key ID matches
	hash := sha256.Sum256(publicKey)
	expectedKeyID := hex.EncodeToString(hash[:8])
	if attestation.KeyID != expectedKeyID {
		return fmt.Errorf("key ID mismatch: expected %s, got %s", expectedKeyID, attestation.KeyID)
	}

	v := NewVerifier()
	v.publicKeys[attestation.KeyID] = ed25519.PublicKey(publicKey)
	return v.Verify(attestation)
}

// ExtractAttestationID extracts the attestation ID from an attestation URI.
// Supports both URL format (.well-known/attestations/{id}.json) and URN format.
func ExtractAttestationID(uri string) (string, error) {
	// Try URL format: https://example.com/.well-known/attestations/{id}.json
	urlPattern := regexp.MustCompile(`/([a-f0-9]{32})\.json$`)
	if matches := urlPattern.FindStringSubmatch(uri); len(matches) == 2 {
		return matches[1], nil
	}

	// Try URN format: urn:recourse:attestation:{id}
	const urnPrefix = "urn:recourse:attestation:"
	if strings.HasPrefix(uri, urnPrefix) && len(uri) >= len(urnPrefix)+32 {
		return uri[len(urnPrefix):], nil
	}

	return "", fmt.Errorf("invalid attestation URI format: %s", uri)
}
